In [ ]:
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch

model_name = "gpt2"

tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

e:\studies_master\deep-learning-in-text-processing\.venv\Lib\site-packages\huggingface_hub\file_download.py:143: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\KOMP\.cache\huggingface\hub\models--gpt2. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

In [2]:
prompt = "In the future, artificial intelligence will"
inputs = tokenizer(prompt, return_tensors="pt")

In [3]:
greedy = model.generate(**inputs, max_length=50)
print("Greedy:\n", tokenizer.decode(greedy[0], skip_special_tokens=True))

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


Greedy:
 In the future, artificial intelligence will be able to do things like search for information about people, and to do things like search for information about people.

"We're going to see a lot of things that are going to be very interesting in


In [ ]:
beam = model.generate(
    **inputs,
    max_length=50,
    num_beams=5,
    early_stopping=True
)
print("Beam search:\n", tokenizer.decode(beam[0], skip_special_tokens=True))

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


Beam search:
 In the future, artificial intelligence will be able to do a lot of things that humans can't. It will be able to do a lot of things that humans can't do. It will be able to do a lot of things that humans can't


In [ ]:
top_k = model.generate(
    **inputs,
    max_length=50,
    do_sample=True,
    top_k=50
)
print("Top-k:\n", tokenizer.decode(top_k[0], skip_special_tokens=True))

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


Top-k:
 In the future, artificial intelligence will change the face of medicine. People will become smarter and cheaper, and they'll spend more time learning about the world around them.

Dr. Christopher A. D. Williams is Executive Director of the California Department


In [ ]:
top_p = model.generate(
    **inputs,
    max_length=50,
    do_sample=True,
    top_p=0.9
)
print("Top-p:\n", tokenizer.decode(top_p[0], skip_special_tokens=True))

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


Top-p:
 In the future, artificial intelligence will help us find, and eliminate dangerous people, and we will not have to worry about any person in the near future, or for the long term.


In [ ]:
texts = {
    "Greedy": tokenizer.decode(greedy[0], skip_special_tokens=True),
    "Beam search": tokenizer.decode(beam[0], skip_special_tokens=True),
    "Top-k": tokenizer.decode(top_k[0], skip_special_tokens=True),
    "Top-p": tokenizer.decode(top_p[0], skip_special_tokens=True)
}

for name, txt in texts.items():
    print("\n=== {} ===".format(name))
    print(f"Length: {len(txt.split())} words")
    print(txt)


=== Greedy ===
Length: 43 words
In the future, artificial intelligence will be able to do things like search for information about people, and to do things like search for information about people.

"We're going to see a lot of things that are going to be very interesting in

=== Beam search ===
Length: 44 words
In the future, artificial intelligence will be able to do a lot of things that humans can't. It will be able to do a lot of things that humans can't do. It will be able to do a lot of things that humans can't

=== Top-k ===
Length: 40 words
In the future, artificial intelligence will change the face of medicine. People will become smarter and cheaper, and they'll spend more time learning about the world around them.

Dr. Christopher A. D. Williams is Executive Director of the California Department

=== Top-p ===
Length: 32 words
In the future, artificial intelligence will help us find, and eliminate dangerous people, and we will not have to worry about any person in the near f

Greedy → spójne, ale mało kreatywne, czasem zapętla się.

Beam search → bardziej spójne, logiczne, ale mało różnorodne.

Top-k → bardziej kreatywne, czasem mniej sensowne.

Top-p → najlepszy balans kreatywności i sensowności.

In [ ]:
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM

sum_model_name = "t5-small"

sum_tokenizer = AutoTokenizer.from_pretrained(sum_model_name)
sum_model = AutoModelForSeq2SeqLM.from_pretrained(sum_model_name)

tokenizer_config.json:   0%|          | 0.00/2.32k [00:00<?, ?B/s]

e:\studies_master\deep-learning-in-text-processing\.venv\Lib\site-packages\huggingface_hub\file_download.py:143: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\KOMP\.cache\huggingface\hub\models--t5-small. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


spiece.model:   0%|          | 0.00/792k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.39M [00:00<?, ?B/s]

config.json:   0%|          | 0.00/1.21k [00:00<?, ?B/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


model.safetensors:   0%|          | 0.00/242M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/147 [00:00<?, ?B/s]

In [9]:
text = """
On November 29th, Lachlan Davidson reported a security vulnerability in React that allows unauthenticated remote code execution by exploiting a flaw in how React decodes payloads sent to React Server Function endpoints.
Even if your app does not implement any React Server Function endpoints it may still be vulnerable if your app supports React Server Components.
This vulnerability was disclosed as CVE-2025-55182 and is rated CVSS 10.0.
The vulnerability is present in versions 19.0, 19.1.0, 19.1.1, and 19.2.0 of:
react-server-dom-webpack
react-server-dom-parcel
react-server-dom-turbopack
Immediate Action Required 
A fix was introduced in versions 19.0.1, 19.1.2, and 19.2.1. If you are using any of the above packages please upgrade to any of the fixed versions immediately.
If your app’s React code does not use a server, your app is not affected by this vulnerability. If your app does not use a framework, bundler, or bundler plugin that supports React Server Components, your app is not affected by this vulnerability.
Affected frameworks and bundlers 
Some React frameworks and bundlers depended on, had peer dependencies for, or included the vulnerable React packages. The following React frameworks & bundlers are affected: next, react-router, waku, @parcel/rsc, @vitejs/plugin-rsc, and rwsdk.
We will update this post with upgrade instructions on how to upgrade as they become available.
Hosting Provider Mitigations 
We have worked with a number of hosting providers to apply temporary mitigations.
You should not depend on these to secure your app, and still update immediately.
Vulnerability overview 
React Server Functions allow a client to call a function on a server. React provides integration points and tools that frameworks and bundlers use to help React code run on both the client and the server. React translates requests on the client into HTTP requests which are forwarded to a server. On the server, React translates the HTTP request into a function call and returns the needed data to the client.

An unauthenticated attacker could craft a malicious HTTP request to any Server Function endpoint that, when deserialized by React, achieves remote code execution on the server. Further details of the vulnerability will be provided after the rollout of the fix is complete.
"""
#source: https://react.dev/blog/2025/12/03/critical-security-vulnerability-in-react-server-components

In [11]:
inputs = sum_tokenizer("summarize: " + text, return_tensors="pt", max_length=512, truncation=True)

summary_ids = sum_model.generate(
    **inputs,
    num_beams=4,
    max_length=120,
    early_stopping=True
)

summary = sum_tokenizer.decode(summary_ids[0], skip_special_tokens=True)
print("SUMMARY:\n", summary)

SUMMARY:
 vulnerability is present in versions 19.0, 19.1.0, 19.1.1, and 19.2.0 of: react-server-dom-webpack react-server-dom-parcel react-server-dom-turbopack Immediate Action Required A fix was introduced in versions 19.0.1, 19.1.2, and 19.2.1.


In [15]:
import evaluate

rouge = evaluate.load("rouge")

scores = rouge.compute(
    predictions=[summary],
    references=[text]
)

print(scores)


{'rouge1': np.float64(0.2273781902552204), 'rouge2': np.float64(0.22377622377622378), 'rougeL': np.float64(0.2273781902552204), 'rougeLsum': np.float64(0.2273781902552204)}


### Summary Quality Evaluation

The generated summary captures the most important point from the original text — that several react-server-dom packages contain a critical vulnerability and that fixed versions are available. It’s factually correct and doesn’t introduce distortions.

However, the summary is very brief and leaves out many relevant details, such as the CVSS score, the risk of remote code execution, which frameworks are affected, and how the vulnerability works. This is reflected in the moderate ROUGE scores (~0.22), which show only partial overlap with the source text.

Overall, the summary is accurate but oversimplified. It conveys the essential message but lacks context that would help readers understand the severity and impact of the vulnerability.